From 08f80d0860bf0a93386bdcab02df6bf6ed36db0a Mon Sep 17 00:00:00 2001
From: ChinhPC <chinhpc@fsoft.com.vn>
Date: Wed, 7 Aug 2019 14:23:38 +0700
Subject: [PATCH 203/628] pinctrl: sh-pfc: r8a77470 PFC support

Add PFC support for the r8a77470 SoC V2 including pin groups for
on-chip devices such as PWM, MSIOF, HSCIF, DU, CAN0, SSI.

Signed-off-by: thongsyho <thong.ho.px@rvc.renesas.com>
Signed-off-by: Nguyen Van Linh [FGA.BU13] <LinhNV25@fsoft.com.vn>
Signed-off-by: vietn <vietn@fsoft.com.vn>
Signed-off-by: Nguyen Van Tu [FGA.AIS] <TuNV16@fsoft.com.vn>
Signed-off-by: ChinhPC <chinhpc@fsoft.com.vn>
---
 drivers/pinctrl/sh-pfc/pfc-r8a77470.c | 276 +++++++++++++++++++++++++++++-----
 1 file changed, 239 insertions(+), 37 deletions(-)

diff --git a/drivers/pinctrl/sh-pfc/pfc-r8a77470.c b/drivers/pinctrl/sh-pfc/pfc-r8a77470.c
index 395d1fd..14aebc9 100644
--- a/drivers/pinctrl/sh-pfc/pfc-r8a77470.c
+++ b/drivers/pinctrl/sh-pfc/pfc-r8a77470.c
@@ -12,37 +12,12 @@
 #define PINMUX_IPSR_GPSR(ipsr, fn)					\
 	PINMUX_DATA(fn##_MARK, FN_##fn, FN_##ipsr)
 
-#define PORT_GP_17(bank, fn, sfx)					\
-	PORT_GP_1(bank, 0,  fn, sfx), PORT_GP_1(bank, 1,  fn, sfx),	\
-	PORT_GP_1(bank, 2,  fn, sfx), PORT_GP_1(bank, 3,  fn, sfx),	\
-	PORT_GP_1(bank, 4,  fn, sfx), PORT_GP_1(bank, 5,  fn, sfx),	\
-	PORT_GP_1(bank, 6,  fn, sfx), PORT_GP_1(bank, 7,  fn, sfx),	\
-	PORT_GP_1(bank, 8,  fn, sfx), PORT_GP_1(bank, 9,  fn, sfx),	\
-	PORT_GP_1(bank, 10, fn, sfx), PORT_GP_1(bank, 11, fn, sfx),	\
-	PORT_GP_1(bank, 12, fn, sfx), PORT_GP_1(bank, 13, fn, sfx),	\
-	PORT_GP_1(bank, 14, fn, sfx), PORT_GP_1(bank, 15, fn, sfx),	\
-	PORT_GP_1(bank, 16, fn, sfx)
-
-#define PORT_GP_23(bank, fn, sfx)					\
-	PORT_GP_17(bank, fn, sfx),					\
-	PORT_GP_1(bank, 17, fn, sfx), PORT_GP_1(bank, 18, fn, sfx),	\
-	PORT_GP_1(bank, 19, fn, sfx), PORT_GP_1(bank, 20, fn, sfx),	\
-	PORT_GP_1(bank, 21, fn, sfx), PORT_GP_1(bank, 22, fn, sfx)
-
-#define PORT_GP_26(bank, fn, sfx)					\
-	PORT_GP_23(bank, fn, sfx),					\
-	PORT_GP_1(bank, 23, fn, sfx), PORT_GP_1(bank, 24, fn, sfx),	\
-	PORT_GP_1(bank, 25, fn, sfx)
-
 #define CPU_ALL_PORT(fn, sfx)						\
-	PORT_GP_23(0, fn, sfx),						\
-	PORT_GP_23(1, fn, sfx),						\
+	PORT_GP_32(0, fn, sfx),						\
+	PORT_GP_32(1, fn, sfx),						\
 	PORT_GP_32(2, fn, sfx),						\
-	PORT_GP_17(3, fn, sfx),						\
-	PORT_GP_1(3, 27, fn, sfx),					\
-	PORT_GP_1(3, 28, fn, sfx),					\
-	PORT_GP_1(3, 29, fn, sfx),					\
-	PORT_GP_26(4, fn, sfx),						\
+	PORT_GP_32(3, fn, sfx),						\
+	PORT_GP_32(4, fn, sfx),						\
 	PORT_GP_32(5, fn, sfx)
 
 enum {
@@ -133,8 +108,10 @@ enum {
 	FN_D13,	FN_MSIOF2_SYNC_A, FN_RX4_C,
 
 	/* IPSR3 */
-	FN_D14, FN_MSIOF2_SS1, FN_TX4_C, FN_CAN1_RX_B, FN_AVB_AVTP_CAPTURE_A,
-	FN_D15, FN_MSIOF2_SS2, FN_PWM4_A, FN_CAN1_TX_B, FN_IRQ2, FN_AVB_AVTP_MATCH_A,
+	FN_D14, FN_MSIOF2_SS1, FN_TX4_C, FN_CAN1_RX_B,
+	FN_AVB_AVTP_CAPTURE_A,
+	FN_D15, FN_MSIOF2_SS2, FN_PWM4_A, FN_CAN1_TX_B,
+	FN_IRQ2, FN_AVB_AVTP_MATCH_A,
 	FN_QSPI0_SPCLK, FN_WE0_N,
 	FN_QSPI0_MOSI_QSPI0_IO0, FN_BS_N,
 	FN_QSPI0_MISO_QSPI0_IO1, FN_RD_WR_N,
@@ -346,9 +323,11 @@ enum {
 	/* IPSR1 */
 	MMC0_D4_MARK, SD1_CD_MARK,
 	MMC0_D5_MARK, SD1_WP_MARK,
-	D0_MARK, SCL3_B_MARK, RX5_B_MARK, IRQ4_MARK, MSIOF2_RXD_C_MARK,	SSI_SDATA5_B_MARK,
+	D0_MARK, SCL3_B_MARK, RX5_B_MARK, IRQ4_MARK,
+	MSIOF2_RXD_C_MARK, SSI_SDATA5_B_MARK,
 	D1_MARK, SDA3_B_MARK, TX5_B_MARK, MSIOF2_TXD_C_MARK, SSI_WS5_B_MARK,
-	D2_MARK, RX4_B_MARK, SCL0_D_MARK, PWM1_C_MARK, MSIOF2_SCK_C_MARK, SSI_SCK5_B_MARK,
+	D2_MARK, RX4_B_MARK, SCL0_D_MARK, PWM1_C_MARK, MSIOF2_SCK_C_MARK,
+	SSI_SCK5_B_MARK,
 	D3_MARK, TX4_B_MARK, SDA0_D_MARK, PWM0_A_MARK, MSIOF2_SYNC_C_MARK,
 	D4_MARK, IRQ3_MARK, TCLK1_A_MARK, PWM6_C_MARK,
 	D5_MARK, HRX2_MARK, SCL1_B_MARK, PWM2_C_MARK, TCLK2_B_MARK,
@@ -364,8 +343,10 @@ enum {
 	D13_MARK, MSIOF2_SYNC_A_MARK, RX4_C_MARK,
 
 	/* IPSR3 */
-	D14_MARK, MSIOF2_SS1_MARK, TX4_C_MARK, CAN1_RX_B_MARK, AVB_AVTP_CAPTURE_A_MARK,
-	D15_MARK, MSIOF2_SS2_MARK, PWM4_A_MARK, CAN1_TX_B_MARK, IRQ2_MARK, AVB_AVTP_MATCH_A_MARK,
+	D14_MARK, MSIOF2_SS1_MARK, TX4_C_MARK, CAN1_RX_B_MARK,
+	AVB_AVTP_CAPTURE_A_MARK,
+	D15_MARK, MSIOF2_SS2_MARK, PWM4_A_MARK, CAN1_TX_B_MARK,
+	IRQ2_MARK, AVB_AVTP_MATCH_A_MARK,
 	QSPI0_SPCLK_MARK, WE0_N_MARK,
 	QSPI0_MOSI_QSPI0_IO0_MARK, BS_N_MARK,
 	QSPI0_MISO_QSPI0_IO1_MARK, RD_WR_N_MARK,
@@ -1282,10 +1263,11 @@ static const unsigned int du0_clk_in_mux[] = {
 	DU0_DOTCLKIN_MARK
 };
 static const unsigned int du0_sync_pins[] = {
-	/* EXVSYNC/VSYNC, EXHSYNC/HSYNC */
-	RCAR_GP_PIN(2, 28), RCAR_GP_PIN(2, 27),
+	/* EXVSYNC/VSYNC, EXHSYNC/HSYNC, EXDISP/EXODDF/EXCDE */
+	RCAR_GP_PIN(2, 29), RCAR_GP_PIN(2, 28), RCAR_GP_PIN(2, 27),
 };
 static const unsigned int du0_sync_mux[] = {
+	DU0_EXODDF_DU0_ODDF_DISP_CDE_MARK,
 	DU0_EXVSYNC_DU0_VSYNC_MARK, DU0_EXHSYNC_DU0_HSYNC_MARK
 };
 static const unsigned int du0_oddf_pins[] = {
@@ -1395,6 +1377,43 @@ static const unsigned int du1_disp_pins[] = {
 static const unsigned int du1_disp_mux[] = {
 	DU1_DISP_MARK
 };
+/* - HSCIF1 ----------------------------------------------------------------- */
+static const unsigned int hscif1_data_pins[] = {
+	/* RX, TX */
+	RCAR_GP_PIN(4, 10), RCAR_GP_PIN(4, 11),
+};
+static const unsigned int hscif1_data_mux[] = {
+	HRX1_A_MARK, HTX1_A_MARK,
+};
+static const unsigned int hscif1_ctrl_a_pins[] = {
+	/* RTS, CTS */
+	RCAR_GP_PIN(4, 13), RCAR_GP_PIN(4, 12),
+};
+static const unsigned int hscif1_ctrl_a_mux[] = {
+	HRTS1_N_A_MARK, HCTS1_N_A_MARK,
+};
+/* - HSCIF2 ----------------------------------------------------------------- */
+static const unsigned int hscif2_data_pins[] = {
+	/* RX, TX */
+	RCAR_GP_PIN(1, 5), RCAR_GP_PIN(1, 6),
+};
+static const unsigned int hscif2_data_mux[] = {
+	HRX2_MARK, HTX2_MARK,
+};
+static const unsigned int hscif2_clk_pins[] = {
+	/* SCK */
+	RCAR_GP_PIN(1, 7),
+};
+static const unsigned int hscif2_clk_mux[] = {
+	HSCK2_MARK,
+};
+static const unsigned int hscif2_ctrl_pins[] = {
+	/* RTS, CTS */
+	RCAR_GP_PIN(1, 9), RCAR_GP_PIN(1, 8),
+};
+static const unsigned int hscif2_ctrl_mux[] = {
+	HRTS2_N_MARK, HCTS2_N_MARK,
+};
 /* - I2C0 ------------------------------------------------------------------- */
 static const unsigned int i2c0_a_pins[] = {
 	/* SCL, SDA */
@@ -1605,6 +1624,42 @@ static const unsigned int mmc_ctrl_pins[] = {
 static const unsigned int mmc_ctrl_mux[] = {
 	MMC0_CLK_SDHI1_CLK_MARK, MMC0_CMD_SDHI1_CMD_MARK,
 };
+/* - MSIOF2 ----------------------------------------------------------------- */
+static const unsigned int msiof2_clk_a_pins[] = {
+	/* SCK */
+	RCAR_GP_PIN(1, 12),
+};
+static const unsigned int msiof2_clk_a_mux[] = {
+	MSIOF2_SCK_A_MARK,
+};
+static const unsigned int msiof2_sync_a_pins[] = {
+	/* SYNC */
+	RCAR_GP_PIN(1, 13),
+};
+static const unsigned int msiof2_sync_a_mux[] = {
+	MSIOF2_SYNC_A_MARK,
+};
+static const unsigned int msiof2_ss1_a_pins[] = {
+	/* SS1 */
+	RCAR_GP_PIN(1, 14),
+};
+static const unsigned int msiof2_ss1_a_mux[] = {
+	MSIOF2_SS1_MARK,
+};
+static const unsigned int msiof2_rx_a_pins[] = {
+	/* RXD */
+	RCAR_GP_PIN(1, 10),
+};
+static const unsigned int msiof2_rx_a_mux[] = {
+	MSIOF2_RXD_A_MARK,
+};
+static const unsigned int msiof2_tx_a_pins[] = {
+	/* TXD */
+	RCAR_GP_PIN(1, 11),
+};
+static const unsigned int msiof2_tx_a_mux[] = {
+	MSIOF2_TXD_A_MARK,
+};
 /* - QSPI ------------------------------------------------------------------- */
 static const unsigned int qspi0_ctrl_pins[] = {
 	/* SPCLK, SSL */
@@ -1927,6 +1982,49 @@ static const unsigned int sdhi2_wp_pins[] = {
 static const unsigned int sdhi2_wp_mux[] = {
 	SD2_WP_MARK,
 };
+/* - SSI -------------------------------------------------------------------- */
+static const unsigned int ssi0_b_data_pins[] = {
+	/* SDATA0 */
+	RCAR_GP_PIN(0, 10),
+};
+static const unsigned int ssi0_b_data_mux[] = {
+	SSI_SDATA0_B_MARK,
+};
+static const unsigned int ssi0129_b_ctrl_pins[] = {
+	/* SCK, WS */
+	RCAR_GP_PIN(0, 8), RCAR_GP_PIN(0, 9),
+};
+static const unsigned int ssi0129_b_ctrl_mux[] = {
+	SSI_SCK0129_B_MARK, SSI_WS0129_B_MARK,
+};
+static const unsigned int ssi1_c_data_pins[] = {
+	/* SDATA1 */
+	RCAR_GP_PIN(0, 7),
+};
+static const unsigned int ssi1_c_data_mux[] = {
+	SSI_SDATA1_C_MARK,
+};
+static const unsigned int ssi1_c_ctrl_pins[] = {
+	/* SCK, WS */
+	RCAR_GP_PIN(0, 5), RCAR_GP_PIN(0, 6),
+};
+static const unsigned int ssi1_c_ctrl_mux[] = {
+	SSI_SCK1_C_MARK, SSI_WS1_C_MARK,
+};
+static const unsigned int ssi6_a_pins[] = {
+	/* SDATA6, SCK, WS */
+	RCAR_GP_PIN(5, 5), RCAR_GP_PIN(5, 3), RCAR_GP_PIN(5, 4),
+};
+static const unsigned int ssi6_a_mux[] = {
+	SSI_SDATA6_A_MARK, SSI_SCK6_A_MARK, SSI_WS6_A_MARK,
+};
+static const unsigned int ssi9_b_data_pins[] = {
+	/* SDATA9 */
+	RCAR_GP_PIN(4, 21),
+};
+static const unsigned int ssi9_b_data_mux[] = {
+	SSI_SDATA9_B_MARK,
+};
 /* - USB0 ------------------------------------------------------------------- */
 static const unsigned int usb0_pins[] = {
 	RCAR_GP_PIN(0, 0), /* PWEN */
@@ -1945,6 +2043,38 @@ static const unsigned int usb1_mux[] = {
 	USB1_PWEN_MARK,
 	USB1_OVC_MARK,
 };
+/* - CAN0 -------------------------------------------------------------------- */
+static const unsigned int can0_data_a_pins[] = {
+		/* CAN0 RXD */
+		RCAR_GP_PIN(0, 11),
+		/* CAN0 TXD */
+		RCAR_GP_PIN(0, 12),
+};
+static const unsigned int can0_data_a_mux[] = {
+		CAN0_RX_A_MARK,
+		CAN0_TX_A_MARK,
+};
+/* - PWM0 -------------------------------------------------------------------- */
+static const unsigned int pwm0_c_pins[] = {
+		RCAR_GP_PIN(5, 11),
+};
+static const unsigned int pwm0_c_mux[] = {
+		PWM0_C_MARK,
+};
+/* - PWM2 -------------------------------------------------------------------- */
+static const unsigned int pwm2_d_pins[] = {
+		RCAR_GP_PIN(5, 27),
+};
+static const unsigned int pwm2_d_mux[] = {
+		PWM2_D_MARK,
+};
+/* - PWM6 -------------------------------------------------------------------- */
+static const unsigned int pwm6_c_pins[] = {
+		RCAR_GP_PIN(1, 4),
+};
+static const unsigned int pwm6_c_mux[] = {
+		PWM6_C_MARK,
+};
 /* - VIN0 ------------------------------------------------------------------- */
 static const union vin_data vin0_data_pins = {
 	.data24 = {
@@ -2118,6 +2248,11 @@ static const struct sh_pfc_pin_group pinmux_groups[] = {
 	SH_PFC_PIN_GROUP(du1_oddf),
 	SH_PFC_PIN_GROUP(du1_cde),
 	SH_PFC_PIN_GROUP(du1_disp),
+	SH_PFC_PIN_GROUP(hscif1_data),
+	SH_PFC_PIN_GROUP(hscif1_ctrl_a),
+	SH_PFC_PIN_GROUP(hscif2_data),
+	SH_PFC_PIN_GROUP(hscif2_clk),
+	SH_PFC_PIN_GROUP(hscif2_ctrl),
 	SH_PFC_PIN_GROUP(i2c0_a),
 	SH_PFC_PIN_GROUP(i2c0_b),
 	SH_PFC_PIN_GROUP(i2c0_c),
@@ -2146,6 +2281,11 @@ static const struct sh_pfc_pin_group pinmux_groups[] = {
 	SH_PFC_PIN_GROUP(mmc_data4),
 	SH_PFC_PIN_GROUP(mmc_data8),
 	SH_PFC_PIN_GROUP(mmc_ctrl),
+	SH_PFC_PIN_GROUP(msiof2_clk_a),
+	SH_PFC_PIN_GROUP(msiof2_sync_a),
+	SH_PFC_PIN_GROUP(msiof2_ss1_a),
+	SH_PFC_PIN_GROUP(msiof2_rx_a),
+	SH_PFC_PIN_GROUP(msiof2_tx_a),
 	SH_PFC_PIN_GROUP(qspi0_ctrl),
 	SH_PFC_PIN_GROUP(qspi0_data2),
 	SH_PFC_PIN_GROUP(qspi0_data4),
@@ -2190,8 +2330,18 @@ static const struct sh_pfc_pin_group pinmux_groups[] = {
 	SH_PFC_PIN_GROUP(sdhi2_ctrl),
 	SH_PFC_PIN_GROUP(sdhi2_cd),
 	SH_PFC_PIN_GROUP(sdhi2_wp),
+	SH_PFC_PIN_GROUP(ssi0_b_data),
+	SH_PFC_PIN_GROUP(ssi0129_b_ctrl),
+	SH_PFC_PIN_GROUP(ssi1_c_data),
+	SH_PFC_PIN_GROUP(ssi1_c_ctrl),
+	SH_PFC_PIN_GROUP(ssi6_a),
+	SH_PFC_PIN_GROUP(ssi9_b_data),
 	SH_PFC_PIN_GROUP(usb0),
 	SH_PFC_PIN_GROUP(usb1),
+	SH_PFC_PIN_GROUP(can0_data_a),
+	SH_PFC_PIN_GROUP(pwm0_c),
+	SH_PFC_PIN_GROUP(pwm2_d),
+	SH_PFC_PIN_GROUP(pwm6_c),
 	VIN_DATA_PIN_GROUP(vin0_data, 24),
 	VIN_DATA_PIN_GROUP(vin0_data, 20),
 	SH_PFC_PIN_GROUP(vin0_data18),
@@ -2252,6 +2402,17 @@ static const char * const du1_groups[] = {
 	"du1_disp",
 };
 
+static const char * const hscif1_groups[] = {
+	"hscif1_data",
+	"hscif1_ctrl_a",
+};
+
+static const char * const hscif2_groups[] = {
+	"hscif2_data",
+	"hscif2_clk",
+	"hscif2_ctrl",
+};
+
 static const char * const i2c0_groups[] = {
 	"i2c0_a",
 	"i2c0_b",
@@ -2298,6 +2459,14 @@ static const char * const mmc_groups[] = {
 	"mmc_ctrl",
 };
 
+static const char * const msiof2_groups[] = {
+	"msiof2_clk_a",
+	"msiof2_sync_a",
+	"msiof2_ss1_a",
+	"msiof2_rx_a",
+	"msiof2_tx_a",
+};
+
 static const char * const qspi0_groups[] = {
 	"qspi0_ctrl",
 	"qspi0_data2",
@@ -2372,6 +2541,15 @@ static const char * const sdhi2_groups[] = {
 	"sdhi2_wp",
 };
 
+static const char * const ssi_groups[] = {
+	"ssi0_b_data",
+	"ssi0129_b_ctrl",
+	"ssi1_c_data",
+	"ssi1_c_ctrl",
+	"ssi6_a",
+	"ssi9_b_data",
+};
+
 static const char * const usb0_groups[] = {
 	"usb0",
 };
@@ -2380,6 +2558,22 @@ static const char * const usb1_groups[] = {
 	"usb1",
 };
 
+static const char * const can0_groups[] = {
+	"can0_data_a",
+};
+
+static const char * const pwm0_groups[] = {
+	"pwm0_c",
+};
+
+static const char * const pwm2_groups[] = {
+	"pwm2_d",
+};
+
+static const char * const pwm6_groups[] = {
+	"pwm6_c",
+};
+
 static const char * const vin0_groups[] = {
 	"vin0_data24",
 	"vin0_data20",
@@ -2408,12 +2602,15 @@ static const struct sh_pfc_function pinmux_functions[] = {
 	SH_PFC_FUNCTION(avb),
 	SH_PFC_FUNCTION(du0),
 	SH_PFC_FUNCTION(du1),
+	SH_PFC_FUNCTION(hscif1),
+	SH_PFC_FUNCTION(hscif2),
 	SH_PFC_FUNCTION(i2c0),
 	SH_PFC_FUNCTION(i2c1),
 	SH_PFC_FUNCTION(i2c2),
 	SH_PFC_FUNCTION(i2c3),
 	SH_PFC_FUNCTION(i2c4),
 	SH_PFC_FUNCTION(mmc),
+	SH_PFC_FUNCTION(msiof2),
 	SH_PFC_FUNCTION(qspi0),
 	SH_PFC_FUNCTION(qspi1),
 	SH_PFC_FUNCTION(scif0),
@@ -2424,8 +2621,13 @@ static const struct sh_pfc_function pinmux_functions[] = {
 	SH_PFC_FUNCTION(scif5),
 	SH_PFC_FUNCTION(scif_clk),
 	SH_PFC_FUNCTION(sdhi2),
+	SH_PFC_FUNCTION(ssi),
 	SH_PFC_FUNCTION(usb0),
 	SH_PFC_FUNCTION(usb1),
+	SH_PFC_FUNCTION(can0),
+	SH_PFC_FUNCTION(pwm0),
+	SH_PFC_FUNCTION(pwm2),
+	SH_PFC_FUNCTION(pwm6),
 	SH_PFC_FUNCTION(vin0),
 	SH_PFC_FUNCTION(vin1),
 };
-- 
2.7.4

